### ISSUE TITLE
Truffle::Interop.respond_to? unusual behavior

### ISSUE THREAD

**Issue by MaxLap** [oracle/truffleruby#1491]
Don't ask me how I land on those things... [deep-cover](https://github.com/deep-cover/deep-cover) needs to do many things to work around the lack of a hook for `require`...

It seems `Truffle::Interop` defines and uses `respond_to?` in a way that is different to the normal `respond_to?`. 

```ruby
Truffle::Interop.respond_to?(:hi)
 #=> ArgumentError: wrong number of arguments (1 for 2)
```

May I suggest renaming the method to something else to avoid the conflict?

https://github.com/oracle/truffleruby/blob/c4cda962fa9f84d524a6216218cf59fbf7bd77be/src/main/ruby/core/truffle/interop.rb#L243

**Comment by marcandre** [comment]
Or change the interface to `def self.respond_to?(object = self, name)`

**Comment by chrisseaton** [comment]
Wow yes I can see how this would be a problem... very curious how you've found this though. You're not using `Truffle::Interop` are you? It's not a supported API. It's just because you're instrumenting all loaded code, is that right?

**Comment by MaxLap** [comment]
It's worse that that.

In order to handle `#autoload` properly and work around implementation details of MRI and JRuby, I need to track all autoloads to deactivate them when they get required/autoloaded. So I start having to iterate over every `Module` to store all existing autoloads. To do this, i use the `constants` method. But in JRuby, there is a case of a Module without that method, so I use `respond_to?` first.

**Comment by pitr-ch** [comment]
Until we fix this here is a workaround:
```ruby
module M
  def self.respond_to?(*args)
    raise
  end
end

respond_to_method = Kernel.method :respond_to?
p respond_to_method.unbind.bind(M).call :to_s #=> true
```

**Comment by MaxLap** [comment]
Thank you, I already have a workaround, but that is a nice one too!

**Comment by pitr-ch** [comment]
Fixed in c30547e32c2c1c174583b940302dd9944ddae325.
